### PR TITLE
[WIP] Add support for Tezos

### PR DESCRIPTION
**Pull request by rlkelly**
## Description

Skeleton for tezos implementation

## Testing instructions

None

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

File: src/Tezos/Address.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "../HexCoding.h"
+
+#include <TrezorCrypto/sha3.h>
+
+#include <cassert>
+
+using namespace TW::Tezos;
+
+bool Address::isValid(const std::string& string) {
+  // TODO: implement
```
Comment (keefertaylor) [inline src/Tezos/Address.cpp]: Addresses are valid if they begin with `tz1`, `tz2` or `tz3` and are base58Check encoded. 

There's a Base58Check functionality provided in TrustCore which you can use.

Here's some test cases:
- Valid:
tz1eZwq8b5cvE2bPKokatLkVMzkxz24z3Don
tz2Rh3NYeLxrqTuvaZJmaMiVMqCajeXMWtYo
tz3RDC3Jdn4j15J7bBHZd29EUee9gVB1CxD9

Invalid: 
NmH7tmeJUmHcncBDvpr7aJNEBk7rp5zYsB1qt // Invalid prefix, valid checksum
tz1eZwq8b5cvE2bPKokatLkVMzkxz24z3AAAA // Valid prefix, invalid checksum
1tzeZwq8b5cvE2bPKokatLkVMzkxz24zAAAAA  // Invalid prefix, invalid checksum

File: src/Tezos/Address.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+Address::Address(const std::string& string) {
+  // TODO: implement
+}
+
+Address::Address(const std::vector<uint8_t>& data) {
+  // TODO: implement
+}
+
+Address::Address(const PublicKey& publicKey) {
+  // TODO: implement
```
Comment (keefertaylor) [inline src/Tezos/Address.cpp]: Create a public key by taking a blake2bhash with length = 20, prefix with "tz1" (bytes [6, 161, 159])

To display, base58encode the bytes.

File: src/Tezos/AddressChecksum.cpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+#include "../Hash.h"
+#include "../HexCoding.h"
+#include <cctype>
+
+using namespace TW;
+using namespace TW::Tezos;
+
+std::string Tezos::checksumed(const Address& address) {
+  // TODO: implement checksum
```
Comment (keefertaylor) [inline src/Tezos/AddressChecksum.cpp]: Calculate a checksum by taking your prefixed data (tz1 + <whatever>), sha256 hashing it twice, then taking the first four bytes. 

Append this to the end of the data (so you have tz1 + <whatever> + <checksum>) and base58encode it

File: src/Tezos/Address.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#pragma once
+
+#include "../PublicKey.h"
+
+#include <array>
+#include <stdint.h>
+#include <vector>
+
+namespace TW {
+namespace Ethereum {
```
Comment (keefertaylor) [inline src/Tezos/Address.h]: (Remember to change your namespaces :) )

File: src/Tezos/Signer.cpp [hunk 5]
```diff
@@ -0,0 +1,26 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Signer.h"
```
Comment (keefertaylor) [inline src/Tezos/Signer.cpp]: Signing an operations bytes: 

privateKey = <some key>
operationHex = <hex encoded string>
watermarkHex = 03
bytesToSign = hex2bin(watermarkHex + operationHex)
hashedOperation = blake2bhas(bytesToSign, length = 32)
signature = sign(hashedOperation, privateKey)
signatureHex = bintohex(signature)

signedBytesForInjection = signatureHex + operationHex


Test case: 
operationhex=deadbeef

File: src/Tezos/Transaction.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#include <boost/multiprecision/cpp_int.hpp>
+
+namespace TW {
+namespace Tezos {
+
+class Transaction {
+public:
+    typedef boost::multiprecision::uint256_t uint256_t;
+
+    uint256_t nonce;
```
Comment (keefertaylor) [inline src/Tezos/Transaction.h]: Here's a canonical JSON representation of a transaction in the Tezos protocol:  
{
	"amount": "1", /// Amount to send as an integer value. I.E. "1" represents .000001 XTZ
	"counter": "30738", // Address Counter, should be +1 over existing value
	"destination": "tz3WXYtyDUNL91qfiCJtVUX746QpNv5i5ve5", // Address receiving
	"fee": "1272", // Fee for operation - should always be the same 
	"gas_limit": "10100", // Gas limit - should always be the same
	"kind": "transaction", // Hardcoded value representing operation kind
	"source": "tz1XVJ8bZUXs7r5NV8dHvuiBhzECvLRLR3jW", // Address making transaction
	"storage_limit": "257" // Gas limit - should always be the same
}

I need to figure out how to to map these values to a hex encoded string.. will update here when I know more.

Comment (keefertaylor) [reply src/Tezos/Transaction.h]: Here's another test case: 

That transaction above should encode to: 
da8eb4f57f98a647588b47d29483d1edfdbec1428c11609cee0da6e0f27cfc3808000081faa75f741ef614b0e35fcc8c90dfa3b0b95721f80992f001f44e81020100026fde46af0356a0476dae4e4600172dc9309b3aa400

(But I don't know how :))

Comment (keefertaylor) [reply src/Tezos/Transaction.h]: Let branch = BMNY6Jkas7BzKb7wDLCFoQ4YxfYoieU7Xmo1ED3Y9Lo3ZvVGdgW

Base58CheckDecode branch and remove first two bytes (should be [1, 52]). Based on branch above this should be:
[218, 142, 180, 245, 127, 152, 166, 71, 88, 139, 71, 210, 148, 131, 209, 237, 253, 190, 193, 66, 140, 17, 96, 156, 238, 13, 166, 224, 242, 124, 252, 56] -> 
[1, 52, 218, 142, 180, 245, 127, 152, 166, 71, 88, 139, 71, 210, 148, 131, 209, 237, 253, 190, 193, 66, 140, 17, 96, 156, 238, 13, 166, 224, 242, 124, 252, 56]

**Comment by hewigovens** [comment]
track it in separate PRs: https://github.com/TrustWallet/wallet-core/pull/93

**Pull request closed without merging**